test/hotspot/gtest/nmt/test_nmt_buffer_overflow_detection.cpp

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright (c) 2021 SAP SE. All rights reserved.
+ * Copyright (c) 2021, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+#include "precompiled.hpp"
+#include "memory/allocation.hpp"
+#include "runtime/os.hpp"
+#include "services/memTracker.hpp"
+#include "utilities/debug.hpp"
+#include "utilities/ostream.hpp"
+#include "unittest.hpp"
+
+#if INCLUDE_NMT
+
+#define DEFINE_TEST(test_function, expected_assertion_message)                            \
+  TEST_VM_FATAL_ERROR_MSG(NMT, test_function, ".*" expected_assertion_message ".*") {     \
+    if (MemTracker::tracking_level() > NMT_off) {                                         \
+      tty->print_cr("NMT overwrite death test, please ignore subsequent error dump.");    \
+      test_function ();                                                                   \
+    } else {                                                                              \
+      /* overflow detection requires NMT to be on. If off, fake assert. */                \
+      guarantee(false,                                                                    \
+                "fake message ignore this - " expected_assertion_message);                \
+    }                                                                                     \
+  }
+
+///////
+
+static void test_overwrite_front() {
+  address p = (address) os::malloc(1, mtTest);
+  *(p - 1) = 'a';
+  os::free(p);
+}
+
+DEFINE_TEST(test_overwrite_front, "header canary broken")
+
+///////
+
+static void test_overwrite_back() {
+  address p = (address) os::malloc(1, mtTest);
+  *(p + 1) = 'a';
+  os::free(p);
+}
+
+DEFINE_TEST(test_overwrite_back, "footer canary broken")
+
+///////
+
+// this should generate two hex dumps, one with the front header, one with the overwritten
+// portion.
+static void test_overwrite_back_long() {
```

Comment:
I think the test isn't really checking that we get two hex dumps, right?

Reply:
Again, bad wording in the comment. This tests that `MallocHeader::print_block_on_error()` prints a hex dump covering both header and the corruption address, and if both are too far apart, that the dump is split up in two parts.